### PR TITLE
Now able to a container on shub by commit (hash)

### LINKED ISSUE #3287: Cannot pull by commit or image hash with v3
With version 2, it was possible to pull an image by the commit or the md5sum of the image file (as described [here]):

```
$ singularity --version
2.6.1-dist
$ singularity pull --name by-commit.simg shub://vsoch/hello-world@42e1f04ed80217895f8c960bdde6bef4d34fab59
Progress |===================================| 100.0% 
Done. Container is at: /tmp/nm-BCiEn3t/by-commit.simg
$ singularity pull --name by-image-hash.simg shub://vsoch/hello-world@ed9755a0871f04db3e14971bec56a33f
Progress |===================================| 100.0% 
Done. Container is at: /tmp/nm-BCiEn3t/by-image-hash.simg
```

When I try the same commands with version 3, it fails:

```
$ singularity --version
singularity version 3.0.3+ds
$ singularity pull --name by-commit.simg shub://vsoch/hello-world@42e1f04ed80217895f8c960bdde6bef4d34fab59
WARNING: Authentication token file not found : Only pulls of public images will succeed
FATAL:   Failed to get manifest from Shub: The requested manifest was not found in singularity hub
$ singularity pull --name by-image-hash.simg shub://vsoch/hello-world@ed9755a0871f04db3e14971bec56a33f
WARNING: Authentication token file not found : Only pulls of public images will succeed
FATAL:   Failed to get manifest from Shub: The requested manifest was not found in singularity hub
```

I didn't spot any mention of this change in the CHANGELOG.  Is pulling by a commit or image hash still supported?

Thanks in advance.

[here]: https://github.com/singularityhub/singularityhub.github.io/wiki/Deploy#order-of-operations

### PR DESCRIPTION
**Pull request by WestleyK**
## Description of the Pull Request (PR):

This PR will add the ability to pull a image from shub by the commit (hash).

#### Example:

```bash
$ singularity pull shub://vsoch/hello-world@ed9755a0871f04db3e14971bec56a33f
# succeeds
$ singularity pull -F shub://vsoch/hello-world@e279432e6d3962777bb7b5e8d54f30f4347d867e
# succeeds
```

## This fixes or addresses the following GitHub issues:

- Fixes #3287 

<br>

### CODE REVIEW

**Review by cclerget — DISMISSED** [review]
LGTM

**Comment by WestleyK** [comment]
Not ready for merge yet. I still need to do some more testing/research on this...

File: pkg/client/shub/api.go [hunk 1]
```diff
@@ -80,10 +79,10 @@ func getManifest(uri ShubURI, noHTTPS bool) (manifest ShubAPIResponse, err error
 	// Do the request, if status isn't success, return error
 	res, err := httpc.Do(req)
 	if res == nil {
-		return ShubAPIResponse{}, fmt.Errorf("No response received from singularity hub")
+		return ShubAPIResponse{}, fmt.Errorf("no response received from singularity hub")
```
Comment (mem) [inline pkg/client/shub/api.go:82]: this is correct, but has nothing to do with the described change

File: pkg/client/shub/api.go [hunk 2]
```diff
@@ -80,10 +79,10 @@ func getManifest(uri ShubURI, noHTTPS bool) (manifest ShubAPIResponse, err error
...
 	}
 	if res.StatusCode == http.StatusNotFound {
-		return ShubAPIResponse{}, fmt.Errorf("The requested manifest was not found in singularity hub")
+		return ShubAPIResponse{}, fmt.Errorf("the requested manifest was not found in singularity hub")
```
Comment (mem) [inline pkg/client/shub/api.go:85]: this is correct, but has nothing to do with the described change

File: pkg/client/shub/pull.go [hunk 3]
```diff
@@ -50,7 +50,7 @@ func DownloadImage(filePath string, shubRef string, force, noHTTPS bool) (err er
 	// Get the image manifest
 	manifest, err := getManifest(ShubURI, noHTTPS)
 	if err != nil {
-		return fmt.Errorf("Failed to get manifest from Shub: %v", err)
+		return fmt.Errorf("failed to get manifest from Shub: %v", err)
```
Comment (mem) [inline pkg/client/shub/pull.go:53]: this is correct, but has nothing to do with the described change

File: pkg/client/shub/util.go [hunk 4]
```diff
@@ -54,19 +54,16 @@ func shubParseReference(src string) (uri ShubURI, err error) {
 	} else if l < 2 {
 		return ShubURI{}, errors.New("Not a valid Shub reference")
 	}
-
-	// look for an @ and split if it exists
-	if strings.Contains(src, `@`) {
-		refParts = strings.Split(src, `@`)
-		uri.digest = `@` + refParts[1]
+	// look for a : and split if it exists
+	if strings.Contains(src, ":") {
```
Comment (mem) [inline pkg/client/shub/util.go:58]: this is doing the work twice: once to look for ":" and a second time to look for ":" again and split the string.

It's equivalent to something like this:

```
if parts := strings.SplitN(src, ":", 2); len(parts) > 1 {
    src = parts[0]
    url.tag = parts[1]
}
```

File: pkg/client/shub/util.go [hunk 5]
```diff
@@ -54,19 +54,16 @@ func shubParseReference(src string) (uri ShubURI, err error) {
...
+		uri.tag = ":" + refParts[1]
 		src = refParts[0]
 	}
 
-	// look for a : and split if it exists
-	if strings.Contains(src, `:`) {
-		refParts = strings.Split(src, `:`)
-		uri.tag = `:` + refParts[1]
-		src = refParts[0]
+	// look for an @ and split if it exists
```
Comment (mem) [inline pkg/client/shub/util.go:64]: I'm not quite sure if the comment is wrong or the code is wrong.

The code is looking for @ and replacing it with `:commit:`, so foo@bar becomes foo:commit:bar, right?

**Review by mem — CHANGES_REQUESTED** [review]
Please update the comment refering to @; the other changes are nice to have, but not required.

**Comment by ikaneshiro** [comment]
@WestleyK just looked at how 2.6.x was getting the image hash and it looks like it's simply parsing the image URL for it: https://github.com/sylabs/singularity/pull/574/files#diff-ce35084decd6d443a423f4356b93f639R190

E.g. https://www.singularity-hub.org/api/container/vsoch/singularity-hello-world:master
We can see an image url of `"image": "https://www.googleapis.com/download/storage/v1/b/singularityhub-legacy/o/singularityhub-legacy%2Fgithub.com%2Fvsoch%2Fsingularity-hello-world%2F54c13f6fad7d3365b43758d55f607fffc6dfc512%2F7e5f47bb07de3da9105896c1b1a61ea7.img.gz?generation=1508189832021221&alt=media"`

So it just looks for the `2F7e5f47bb07de3da9105896c1b1a61ea7.img.gz` and uses that for the hash. Not sure what hashing alg is used for that but hopefully this will help.

Not 100% sure if there is more complex behavior to the regex but thats what it seems like at first glance.

File: pkg/client/shub/util.go (hunk 5 above)
Comment (WestleyK) [reply pkg/client/shub/util.go:64]: yeah, thats correct. I still need to fix the comments to make sense.

File: pkg/client/shub/api.go (hunk 1 above)
Comment (WestleyK) [reply pkg/client/shub/api.go:82]: When I was changing this file while debugging, I could't help myself, I needed to fix that! But I'll change it back, this file did not need modifying.

File: pkg/client/shub/pull.go (hunk 3 above)
Comment (WestleyK) [reply pkg/client/shub/pull.go:53]: Same story here, I'll revert this file.

File: pkg/client/shub/api.go (hunk 1 above)
Comment (sashayakovtseva) [reply pkg/client/shub/api.go:82]: IMO it is perfectly fine to apply minor improvements as you work on something. See http://www.extremeprogramming.org/rules/refactor.html

File: pkg/client/shub/util.go (hunk 5 above)
Comment (sashayakovtseva) [reply pkg/client/shub/util.go:64]: @WestleyK will you update this PR?

Comment (WestleyK) [reply pkg/client/shub/util.go:64]: @sashayakovtseva, Yes, soon-ish. But there is still some testing/comparing with 2.X before this can be merged.

**Comment by dtrudg** [comment]
This has gone stale - should probably be attacked as described in https://github.com/hpcng/singularity/pull/3652#issuecomment-505041233 if it's going to be attacked further.

**Pull request closed without merging**